### ISSUE TITLE
nan propagates through backward pass even when not accessed

### ISSUE THREAD

**Issue by carlosgmartin** [pytorch/pytorch#15506]
## 🐛 Bug

`nan` propagates through computations during a `backward()` pass even if it wasn't accessed at all.

## To reproduce

```
a = torch.arange(4, dtype=torch.float, requires_grad=True)
a.log()[-1].backward()
print(a.grad)
```
prints `tensor([   nan, 0.0000, 0.0000, 0.3333])`

## Expected behavior

It should print `tensor([0.0000, 0.0000, 0.0000, 0.3333])`, like the following code does:
```
a = torch.arange(4, dtype=torch.float, requires_grad=True)
a[-1].log().backward()
print(a.grad)
```

The same should be true if `[-1]` is replaced with `.max()` (which causes the same entry to be accessed, in this particular example), which is how I originally encountered the issue.

## Environment

```
Collecting environment information...
PyTorch version: 1.0.0a0+4c11dee
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Mac OSX 10.14.2
GCC version: Could not collect
CMake version: version 3.13.0

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 10.0.130
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda/lib/libcudnn.7.dylib
/usr/local/cuda/lib/libcudnn.dylib
/usr/local/cuda/lib/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.0                    <pip>
[conda] mkl                       2018.0.3                      1  
[conda] mkl-include               2019.0                    <pip>
[conda] mkl-service               1.1.2            py36h6b9c3cc_4  
[conda] mkl_fft                   1.0.2            py36h6b9c3cc_0  
[conda] mkl_random                1.0.1            py36h5d10147_1  
[conda] torch                     1.0.0a0+4c11dee           <pip>
[conda] torchfile                 0.1.0                     <pip>
[conda] torchvision               0.2.0                     <pip>
```

**Comment by vishwakftw** [comment]
This behavior is explained in #15131 .

**Comment by carlosgmartin** [comment]
I don’t understand this part:

> any operation that uses the NaN input to compute its grad_in from grad_out (like mul, or pow) will see something 0*NaN and that is most reasonably evaluated as NaN in general

The output is *completely and totally unaffected* by the first element. I don’t see why its gradient should be *anything* other than zero. Having to introduce hooks and manually mess around with indices during backprop is a lot of trouble to get around what is clearly incorrect behavior.

**Comment by vishwakftw** [comment]
Functionally, this is how the graph (forward pass) would look like:
Input -> Index Select -> Log -> Sum

The backwards pass is the same graph, except that the directions are swapped
SumBackward -> LogBackward -> IndexSelectBackward -> InputGrad

I hope you see the issue here:
SumBackward is a tensor of ones, log backward would be the reciprocals of the _entire_ input, index select backward is a mask of 0s and 1s depending on what were indexed and what weren’t.

A composition of these gradients gives the final gradient. The issue is that, the entire tensor is continually passed, and not a slice.

Hope this cleared up any confusion. I would still say that hooks are your best bet.

**Comment by vishwakftw** [comment]
Index select _backwards_ is most probably a mask. You can have a look at `tools/autograd/templates/Functions.cpp` for more details.

**Comment by t-vi** [comment]
The problem is that index select will give zero gradient for the non-selected indices.
But you then multiply by NaN when taking the log derivative.
There is no way to distinguish "0 gradient from not used" to 0 from "the gradient happens to be 0". Then you arrive at "NaN * 0". The NaN in the gradient is inserted by the derivative formula of "log" not of the indexing.
The best way is to do the masking yourself, probably.

**Comment by carlosgmartin** [comment]
Why is selection handled through multiplication by 0, though? Shouldn’t it be handled by a conditional like torch.where?

**Comment by carlosgmartin** [comment]
In either case, shouldn’t this “forced 0” issue be automically handled by PyTorch’s IndexSelectBackward?

**Comment by albanD** [comment]
Hi,
The problem is that you're looking at more than one op at a time. If you see the indexing op as a single one, it takes a big tensor as input and returns a smaller. So the backward takes a small tensor of gradients and returns a big tensor of gradients. Unfortunately, a tensor can only contain numbers. And so the right value for the indexing op is to return 0 gradient in the places that were not used.

Unfortunately here, you expect pytorch to consider 2 ops together (indexing and log) which it cannot do. All the gradients are computed with the chain rule one op at a time.

**Comment by carlosgmartin** [comment]
If I can manually create a hook to fix this incorrect behavior (as in https://github.com/pytorch/pytorch/issues/15131#issuecomment-447149154), is there a reason PyTorch can’t do it automatically?

**Comment by albanD** [comment]
The thing ia that the manual backward hook will use informations from the masking op to change the gradients computes for the log op. So the hook needs to ve aware of these two ops and how they follow each other.
So you can manually create this hook because you know how these ops interact, pytorch cannot as it only looks at one op at a time :/

**Comment by soumith** [comment]
based on the discussion above, I am closing the issue. please reopen if you disagree or see actionable items for pytorch devs to do.
